lib/service/storage/elasticsearch.js

```diff
@@ -2008,6 +2009,102 @@ class ElasticSearch extends Service {
     return response;
   }
 
+  /**
+   * Creates or replaces multiple documents at once.
+   *
+   * @param {String} index - Index name
+   * @param {String} collection - Collection name
+   * @param {Object[]} documents - Documents
+   * @param {Object} options - refresh (undefined), retryOnConflict (0), timeout (undefined), userId (null)
+   *
+   * @returns {Promise.<{ items, errors }>
+   */
+  async mUpsert (
+    index,
+    collection,
+    documents,
+    { refresh, retryOnConflict = 0, timeout, userId=null } = {})
+  {
+    const
+      esIndex = this._getESIndex(index, collection),
+      esRequest = {
+        body: [],
+        refresh,
+        timeout
+      },
+      user = getUserId(userId),
+      now = Date.now(),
+      kuzzleMeta = {
+        doc: {
+          _kuzzle_info: {
+            updatedAt: now,
+            updater: user
+          }
+        },
+        upsert: {
+          _kuzzle_info: {
+            author: user,
+            createdAt: now
+          }
+        }
+      },
+      {
+        rejected,
+        extractedDocuments
+      } = this._extractMDocuments(
+        documents,
+        kuzzleMeta,
+        { 
+          prepareMUpsert: true,
+          requireId: true
+        }
+      );
+
+    assertWellFormedRefresh(esRequest);
+
+    /**
+     * @warning Critical code section
+     *
+     * request can contain more than 10K elements
+     */
+    for (const extractedDocument of extractedDocuments) {
```

Comment:
Critical code section should use `for (let ...)` loop. Also I'm not sure a loop that simply create another array can be considered as critical since it's a very common operation